jax/_src/lax/lax.py

```diff
@@ -1088,19 +1091,29 @@ def argmax(operand: Array, axis: int,
   return argmax_p.bind(operand, axes=(axis,),
                        index_dtype=dtypes.canonicalize_dtype(index_dtype))
 
-def reduce(operand: Array, init_value: Array, computation: Callable,
-           dimensions: Sequence[int]) -> Array:
+def reduce(operands: Array, init_values: Array, computation: Callable,
+           dimensions: Sequence[int], **reduce_kwargs: Any) -> Array:
   """Wraps XLA's `Reduce
   <https://www.tensorflow.org/xla/operation_semantics#reduce>`_
   operator.
   """
-  monoid_reducer = _get_monoid_reducer(computation, init_value)
+  flat_operands, operand_tree = tree_util.tree_flatten(operands)
+  flat_init_values, init_value_tree = tree_util.tree_flatten(init_values)
+  assert operand_tree == init_value_tree, (
+      'Init values must have same pytree structure as operands.')
+  assert len(flat_operands) == len(flat_init_values), (
+      'Must have same number of init values as operands')
+  monoid_reducer = _get_monoid_reducer(computation, flat_init_values, **reduce_kwargs)
   if monoid_reducer:
-    return monoid_reducer(operand, dimensions)
+    return monoid_reducer(*flat_operands, dimensions)
   else:
-    jaxpr, consts = _reduction_jaxpr(computation, _abstractify(init_value))
-    return reduce_p.bind(operand, init_value, computation=computation,
+    init_avals = tree_util.tree_map(_abstractify, init_values)
```

Comment:
Nit: I might be tempted to flatten and then abstractify.

Reply:
Done.